pkg/pulumiyaml/run.go

```diff
@@ -1758,6 +1761,48 @@ func (e *programEvaluator) evaluateBuiltinInvoke(t *ast.InvokeExpr) (interface{}
 	return performInvoke(args)
 }
 
+func (e *programEvaluator) evaluateBuiltinMethodCall(t *ast.MethodExpr) (interface{}, bool) {
+	args, ok := e.evaluateExpr(t.CallArgs)
+	if !ok {
+		return nil, false
+	}
+
+	performCall := e.lift(func(args ...interface{}) (interface{}, bool) {
+		r, ok := e.evaluateExpr(t.ResourceName)
+		if !ok {
+			e.error(t.ResourceName, fmt.Sprintf("unable to resolve resource name: %v", t.ResourceName))
+			return nil, false
+		}
+		res := r.(lateboundResource)
+
+		_, functionName, err := ResolveFunction(e.pkgLoader, t.FuncToken.Value, nil)
+		if err != nil {
+			matched := false
+			if res, ok := r.(*lateboundCustomResourceState); ok {
+				hint := res.pkg.ResourceTypeHint(res.typeTok)
+				for _, m := range hint.Resource.Methods {
+					if m.Name == t.FuncToken.Value {
+						functionName = FunctionTypeToken(m.Function.Token)
+						matched = true
+						break
+					}
+				}
+			}
+			if !matched {
+				return e.error(t, err.Error())
+			}
+		}
```

Comment:
What is this doing? It looks like we are doing type-checking within the apply body.

A comment explaining what is going on here would be appreciated.

Reply:
I think I've moved the type-checking to the proper place, storing a lookup table for resource methods to validate a resource+method